### ISSUE TITLE
[SYCL] clang++: `-fsycl` invokes the compiler front-end even in pure preprocessor mode (`-E`)

### ISSUE THREAD

**Issue by 0x6675636b796f75676974687562** [intel/llvm#3925]
`clang++ -E` is supposed to run just the preprocessor.
The preprocessor mode combined with `-fsycl` (`clang++ -E -fsycl`), however, invokes the compiler front-end.

**To Reproduce**

Consider the following partial (non-compilable) code:

```c++
// test.cc
#if defined(__cplusplus)
Foo->bar = __cplusplus;
#endif
```

Whenever the above code is preprocessed with either

```
g++ -xc++ -g -fpch-preprocess -v -dD -E test.cc 2>/dev/null
```

or

```
clang++ -xc++ -g -fpch-preprocess -v -dD -E test.cc 2>/dev/null
```

both compilers produce their output w/o any error, and the result may be included elsewhere, parsed using any 3rd-party tools, etc.

If your `clang++` fork is used and `CXXFLAGS` include `-fsycl`, the error message is reported:

```
test.cc:2:1: error: unknown type name 'Foo'
Foo->bar = __cplusplus;
^
test.cc:2:4: error: cannot use arrow operator on a type
Foo->bar = __cplusplus;
   ^
2 errors generated.
```

Is this the intended behaviour?

**Environment:**

- OS: Linux (`Linux proxima 4.19.0-9-amd64 #1 SMP Debian 4.19.118-2 (2020-04-29) x86_64 GNU/Linux`)
- Target device and vendor: N/A
- DPC++ version:
    ```
    clang version 13.0.0 (https://github.com/intel/llvm.git 525d253b4d7325eacfddd7284151c2bd207ed304)
    Target: x86_64-unknown-linux-gnu
    Thread model: posix
    ```
- Dependencies version: N/A

**Comment by bader** [comment]
@mdtoguchi, could you take a look, please?

**Comment by mdtoguchi** [comment]
This is expected behavior.  When creating preprocessed files, we need to take into account the integration header.  The integration header is created by the device compilation which will be used to create the preprocessed file for the host compilation.

**Comment by Michoumichmich** [comment]
Maybe you could use `#if defined(__SYCL_DEVICE_ONLY__)` to circumvent your problem?

**Comment by unix-junkie** [comment]
> This is expected behavior. When creating preprocessed files, we need to take into account the integration header. The integration header is created by the device compilation which will be used to create the preprocessed file for the host compilation.

@mdtoguchi, thank you for looking into this!

Still, isn't `-E` supposed to also imply `-fsycl-device-only`?

**Comment by 0x6675636b796f75676974687562** [comment]
> Maybe you could use `#if defined(__SYCL_DEVICE_ONLY__)` to circumvent your problem?

@Michoumichmich, thank you for your suggestion, but, unfortunately, we coudln't.

We accept a user-specified compiler along with its default switches (similarly to the way how _CMake_ accepts `CMAKE_<LANG>_COMPILER` and `CMAKE_<LANG>_COMPILER_ARG1`) and try to infer the compiler information from the preprocessor output.

Looks like our assumption that `-E` meant "do not compile" was wrong.

**Comment by Michoumichmich** [comment]
[This](https://github.com/intel/llvm/commit/3a18db6ab266d3ba4890b0de75c9408c8c5901f0) is the commit that "broke" my toolchain integration fyi

**Comment by mdtoguchi** [comment]
@unix-junkie, unfortunately no, `-E` does not imply `-fsycl-device-only`.  The generated preprocessed file needs to be complete and be able to be fully compiled for both host and device.  If you were to use `-E` with `-fsycl-device-only` there would be no full compile dependency, but the preprocessed information would only be for the device.

**Comment by AlexeySachkov** [comment]
@0x6675636b796f75676974687562, looks like the question was answered, can we close the issue?

**Comment by 0x6675636b796f75676974687562** [comment]
@AlexeySachkov, yes, it can be safely closed now.

Support for Intel's fork of clang as well as Intel DPC++ has been implemented in CLion 2022.1 ([CPP-25963](https://youtrack.jetbrains.com/issue/CPP-25963)), so the issue is no longer relevant.
